### ISSUE TITLE
"all pools" filter

### ISSUE THREAD

**Issue by clarfonthey** [zrepl/zrepl#601]
Right now, it's possible to request backing up all pools by providing the `>` pattern to zrepl, but it's not possible to disable patterns based upon datasets within pools.

Essentially, I recently reorganised my pools such that the entire pools are snapshotted, except for the datasets within a `nosnap` dataset at the root of the pools. It would be helpful to be able to annotate this via zrepl, using (syntax bikesheddable):

```
filesystems:
    "/<": true
    "/nosnap<": false
```

where here, `/` indicates "all pools".

Alternatively, a `*` pattern could be added that represents any pool or dataset. For example, you could do `*/nosnap` to mean "the `nosnap` dataset in any pool" or `pool/*/nosnap` to mean "the `nosnap` dataset within any dataset inside the pool `pool`". This would also extend to `**`, which could match multiple levels of datasets.

Of course, I understand the desire to avoid adding such wildcards, since they could make the lookup logic for zrepl much more complicated; this is why I propose just adding an "all pools" wildcard for now, since it at least helps solve the use case of using the same filesystem conventions on multiple pools, and/or wanting to copy the same config to multiple machines who may have differently named pools.

**Comment by problame** [comment]
We already have a similar request at https://github.com/zrepl/zrepl/issues/287

The unanswered problem there, which I repeat, for you to answer for your feature request:

> I think the difficulty here lies in prioritization compared to the < 'subtree wildcard' and the non-wildcard filter rules.
(see https://zrepl.github.io/configuration/filter_syntax.html for the existing prioritization rules).
> In the current config syntax, the order of the filter rules does not matter and cannot be easily reconstructed due to constraints by the yaml parsing library.
> How would you go about this?

---

> , since they could make the lookup logic for zrepl much more complicated; 

What exactly are you referring to here?
I think figuring out which sender and dataset corresponds to which receiver dataset based on `receiver.root_fs + sender.name <==> sender.name`  could become tricky once we allow `**` style globs, but I don't think it would be a problem for `*` style globs.
What do you think? cc @InsanePrawn 

---

Note that this is much simpler than https://github.com/zrepl/zrepl/issues/253 and almost ready to implement, IMO.

**Comment by clarfonthey** [comment]
So, I think that the most logical way is to go by depth-based specificity, then they're ordered lexicographically. So, from most to least specific, we'd have `path`, `*`, and `**`. When performing ordering, we can treat `path<` as an alias for `path/**`.

There are also a few cases of `**` that should be forbidden if we allow it but aren't necessary to worry about when doing the ordering.

Here's an example ordering, from most to least specific:

* `one/two/three`
* `one/two/*`
* `one/two/**` (`one/two<`)
* `one/*/three`
* `one/*/*`
* `one/*/**` (`one/*<`)
* `one/**/three`
* `one/**/*`
* forbidden: `one/**/**` (equivalent to `one/**`)
* `*/two/three`
* `*/two/*`
* `*/two/**` (`*/two<`)
* `*/*/three`
* `*/*/*`
* `*/*/**`
* `*/**/three`
* forbidden: `*/**/*` (equivalent to `*/*/**`)
* `**/two/three`
* `**/two/*`
* `**/two/**`
* `**/*/three`
* `**/*/*`
* forbidden: `**/*/**` (equivalent to `*/**`)

**Comment by problame** [comment]
>  E.g. if we're a pull job and looking at our local backup destination, we'll assume that what's in the destination is controlled by zrepl, i.e. either a payload dataset or a placeholder. Somewhat exclusive control over the target hierarchy has always been a firm assumption for zrepl AFAIK?

I think we do, for pruning. Datasets that don't have a correspondence on the receiver aren't touched currently. 

https://github.com/zrepl/zrepl/blob/1f0f2f856966713c06b76943d9de0c560f820557/daemon/pruner/pruner.go#L397

If we changed that behavior, then we break the current conflict resolution workflow for filesystem re-creations on the sender.
I.e., right now, when the sender destroys and re-creates a filesystem with the same name, you resolve the conflict manually by ranming the recv side dataset. Often, that would be `${original_name}.old` or similar.

---

@clarfonthey please respond to @InsanePrawn 's other follow-up questions.

**Comment by InsanePrawn** [comment]
> > E.g. if we're a pull job and looking at our local backup destination, we'll assume that what's in the destination is controlled by zrepl, i.e. either a payload dataset or a placeholder. Somewhat exclusive control over the target hierarchy has always been a firm assumption for zrepl AFAIK?
> 
> I think we do, for pruning. Datasets that don't have a correspondence on the receiver aren't touched currently.

Right, forgot about that.

> 
> https://github.com/zrepl/zrepl/blob/1f0f2f856966713c06b76943d9de0c560f820557/daemon/pruner/pruner.go#L397

I don't understand how/why this behavior is changed by adding slightly different wildcard characters to the existing `<`? I think most of the same assumptions, problems and solutions should apply?

**Comment by problame** [comment]
Yes, that's true, unfounded concerns on my side. The reverse mapping concern isn't relevant here because we already got the exact list of sender filesystems.

Let me rephrase the reverse mapping concern:
If we admit `**` wildcards, we won't be able to take recv-side names, and translate them back into send-side names, given just the send-side `filesystems` YAML.

**Comment by problame** [comment]
But I think we don't need that reverse mapping capability anywhere in zrepl today, and I don't think we will in the future. 
So, I would accept a patch that implements `*` and `**` style wildcards, if the remaining concerns are addressed.

**Comment by InsanePrawn** [comment]
> Yes, that's true, unfounded concerns on my side. The reverse mapping concern isn't relevant here because we already got the exact list of sender filesystems.
> 
> Let me rephrase the reverse mapping concern: If we admit `**` wildcards, we won't be able to take recv-side names, and translate them back into send-side names, given just the send-side `filesystems` YAML.

> But I think we don't need that reverse mapping capability anywhere in zrepl today,

I mean... AFAICT we can do that as much as we can for this case nowdays:

```yml
filesystems:
  "pool_a/data<": true
  "pool_b/data<": true
```

given a `root_fs` of `pool_t/sink`, you can either just substract the `root_fs` prefix to have your entire path including a potentially globbed pool name - or (if you for some reason don't trust transmitted pool names) you're already in undecidable territory between the clearly hardcoded pool_a and pool_b nowadays.


> and I don't think we will in the future.

Technically rename detection in the far-out future is gonna need some form of it, but a proper mechanism for that will also not easily be broken by more wildcards AFAICT :)

**Comment by problame** [comment]
>  If someone has an informed opinion on whether "a/b/*": true should also match a/b, I'm open, otherwise I'd assume it does not match.

It should behave like bash globs:

```
[cs@cslap test]$ tree
.
├── a
└── aa

0 directories, 2 files
[cs@cslap test]$ file *
a:  empty
aa: empty
[cs@cslap test]$ file a*
a:  empty
aa: empty
[cs@cslap test]$ file aa*
aa: empty
[cs@cslap test]$ file a/*
a/*: cannot open `a/*' (Not a directory)
[cs@cslap test]$ file */*
*/*: cannot open `*/*' (No such file or directory)

```

**Comment by clarfonthey** [comment]
I agree; to me, `a/b/*` should only match children of `a/b` and not `a/b` itself.

Apologies for not responding earlier, but it appears that all the open questions have been solved… unless I'm missing something?

**Comment by colttt** [comment]
it would also be nice if we can add support for `>` which mean just all its children

**Comment by InsanePrawn** [comment]
> it would also be nice if we can add support for `>` which mean just all its children

As I understand it, your `a>` would then just be an alias of the proposed `a/**` ? (or `a/*` depending on how you meant it)

Personally I'd prefer **not** to extend both wildcard types simultaneously for equivalent meanings.

**Comment by problame** [comment]
Yes, for "just all children", `parent/**` should be it. 
I feel like we need to provide the backwards-compatibility for `<`, but we shouldn't add to the confusion with `>`.
If we don't have backwards-compatibility, but get good error messages for use of `<`, I'd be fine with no backwards-compat as well.
But it seems cheap to do.

**Comment by t-oster** [comment]
Hi, I just want to add a use case:
We use zrepl on our proxmox nodes in order to have automatically snapshots for all vms and containers.
The vms and containers are created and destroyed frequently, so we want to catch all cases just with patterns.

However: For some backup-reasons, proxmox creates some zvols like "vm-108-fleece-0" which are used during backup and then destroyed again. If zrepl creates a snapshot during that backup, proxmox won't be able to destroy the volume, because it has snapshots.

So we would need a '*' wildcard to exclude datasets like '*-fleece-*' from our snapjob.
If this feature is added, please make the '*' work on a name-level, not just as a placeholder for a whole dataset in the tree.
